### PR TITLE
Graph-activity-feed notification in node js for teams, group chat and user scope

### PR DESCRIPTION
**Pull request by NikitaJain-MSFT**
Sample to showcase all scenario for graph-activity-feed notification - Node
It covers the following functionalities:

 Group Chat
 1) Send Notification to user group chat
 2) Send Notification with custom topic to user group chat

Teams
1) Send notification to user in channel
2) Send notification with custom topic in channel
3) Notify user about a channel tab

User
1) Send notification to user
2) Send notification with custom topic

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -211,6 +211,7 @@
 [graph#graphproactiveinstallation#js]:samples/graph-proactive-installation/nodejs
```
Comment (v-chetsh) [inline README.md:212]: Remove empty line

File: samples/graph-activity-feed/nodejs/Manifest/manifest.json [hunk 2]
```diff
@@ -0,0 +1,80 @@
+{
+  "$schema": "https://developer.microsoft.com/en-us/json-schemas/teams/v1.8/MicrosoftTeams.schema.json",
+  "manifestVersion": "1.8",
+  "version": "1.0.1",
+  "id": "<<YOUR-MANIFEST-ID>>",
+  "packageName": "com.microsoft.teams.devapp",
```
Comment (v-chetsh) [inline samples/graph-activity-feed/nodejs/Manifest/manifest.json:6]: Add proper package name as per your application name

File: samples/graph-activity-feed/nodejs/Manifest/manifest.json [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+    "privacyUrl": "https://www.teams.com/privacy",
+    "termsOfUseUrl": "https://www.teams.com/termsofuse"
+  },
+  "icons": {
+    "color": "color.png",
+    "outline": "outline.png"
+  },
+  "name": {
+    "short": "NotifyFeedApp",
+    "full": "NotifyFeedApp for Send Activity Feed "
```
Comment (v-chetsh) [inline samples/graph-activity-feed/nodejs/Manifest/manifest.json:19]: Remove space after Feed

File: samples/graph-activity-feed/nodejs/Manifest/manifest.json [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+        "privateChatTab"
+      ]
+    }
+  ],
+  "permissions": [
+    "identity",
+    "messageTeamMembers"
+  ],
+  "validDomains": [
+    "*.ngrok.io"
```
Comment (v-chetsh) [inline samples/graph-activity-feed/nodejs/Manifest/manifest.json:45]: Add placeholder for domain like <Application Domain>

File: samples/graph-activity-feed/nodejs/.vscode/launch.json [hunk 5]
```diff
@@ -0,0 +1,15 @@
+{
```
Comment (AbhijeetThakur-msft) [inline samples/graph-activity-feed/nodejs/.vscode/launch.json:1]: is this file required?

File: samples/graph-activity-feed/nodejs/Readme.md [hunk 6]
```diff
@@ -0,0 +1,117 @@
+# Notify Feed App
+
+Nodejs Activity Feed sample using Tab.
+
+This sample has been created using [Microsoft Graph](https://docs.microsoft.com/en-us/graph/overview?view=graph-rest-beta), it shows how trigger a Activity feed notification from your Tab, it triggers the feed notification for User, Chat and Team scope and send back to conversation.
```
Comment (AbhijeetThakur-msft) [inline samples/graph-activity-feed/nodejs/Readme.md:5]: here provide link of the sample doc as well

File: samples/graph-activity-feed/nodejs/Readme.md [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+    - `offline_access`
+    - `OpenId`
+    - `profile`
+    - `ChannelMessage.Send`
+    - `ChatMessage.Send`
+    - `Chat.ReadWrite`
+    - `TeamsActivity.Send`
+    - `TeamsAppInstallation.ReadForUser.All`.
+
+**Note** Your need to add `TeamsActivity.Send` as Application level permissions
```
Comment (AbhijeetThakur-msft) [inline samples/graph-activity-feed/nodejs/Readme.md:77]: Rephrase the statement as: You need to add

File: samples/graph-activity-feed/nodejs/Readme.md [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+
+    ```bash
+    npm start
+    ```
+## User Interaction with Tab Activity Feed App
+
+- Install TabActivityFeed manifest in Teams
+- Add Tab in Personal, GroupChat or Team scope
+- Fill the Details in Page and click on Send notification button
+![](Images/image1.png)
```
Comment (AbhijeetThakur-msft) [inline samples/graph-activity-feed/nodejs/Readme.md:109]: Give proper name to image, do not use no.s

File: samples/graph-activity-feed/nodejs/auth.js [hunk 9]
```diff
@@ -0,0 +1,30 @@
+var request = require('request');
+var graph = require('@microsoft/microsoft-graph-client');
+var Q = require('q');
+
+// The auth module object.
+var auth = {};
+auth.getAccessToken = function (tenantId) {
```
Comment (AbhijeetThakur-msft) [inline samples/graph-activity-feed/nodejs/auth.js:7]: Give some comments to this file methods

File: samples/graph-activity-feed/nodejs/static/scripts/auth-delegated.js [hunk 10]
```diff
@@ -0,0 +1,108 @@
...
+                failureCallback: function (error) {
+                    reject("Error getting token: " + error);
+                }
+            });
+
+        });
+    }
+
+    // 2. Exchange that token for a token with the required permissions
+    //    using the web service (see /auth/token handler in app.js)
```
Comment (AbhijeetThakur-msft) [inline samples/graph-activity-feed/nodejs/static/scripts/auth-delegated.js:22]: correct document indentation, mainly for comments

File: samples/graph-activity-feed/nodejs/static/scripts/auth-delegated.js [hunk 11]
```diff
@@ -0,0 +1,108 @@
...
+                    resolve(data);
+                },
+                failureCallback: (reason) => {
+                    reject(JSON.stringify(reason));
+                }
+            });
+        });
+    }
+
+    // In-line code
```
Comment (AbhijeetThakur-msft) [inline samples/graph-activity-feed/nodejs/static/scripts/auth-delegated.js:77]: Add more meaningful comment

File: samples/graph-activity-feed/nodejs/static/scripts/auth-delegated.js [hunk 12]
```diff
@@ -0,0 +1,108 @@
...
+    getClientSideToken()
+        .then((clientSideToken) => {
+            return getServerSideToken(clientSideToken);
+        })
+        .catch((error) => {
+            if (error === "invalid_grant") {
+                console.log(`Error: ${error} - user or admin consent required`);
+                // Display in-line button so user can consent
+                let button = display("Consent", "button");
```
Comment (AbhijeetThakur-msft) [inline samples/graph-activity-feed/nodejs/static/scripts/auth-delegated.js:86]: where are we using this button? It might come as undefined, please test this scenario

File: samples/graph-activity-feed/nodejs/views/GroupChatNotification.ejs [hunk 13]
```diff
@@ -0,0 +1,263 @@
+<!DOCTYPE html>
+<html lang="en">
+<head>
+    <head>
+        <script src="https://statics.teams.cdn.office.net/sdk/v1.6.0/js/MicrosoftTeams.min.js"
+            integrity="sha384-mhp2E+BLMiZLe7rDIzj19WjgXJeI32NkPvrvvZBrMi5IvWup/1NUfS5xuYN5S3VT"
+            crossorigin="anonymous"></script>
+        <script src="https://statics.teams.microsoft.com/sdk/v1.7.0/js/MicrosoftTeams.min.js"></script>
+        <script src="https://unpkg.com/@microsoft/teams-js/dist/MicrosoftTeams.min.js" crossorigin="anonymous"></script>
```
Comment (AbhijeetThakur-msft) [inline samples/graph-activity-feed/nodejs/views/GroupChatNotification.ejs:9]: there 3 same imports for teams sdk import at line 5, please use only one import statement and the latest version of it, please check from docs

File: samples/graph-activity-feed/nodejs/views/GroupChatNotification.ejs [hunk 14]
```diff
@@ -0,0 +1,263 @@
...
+        <script src="https://ajax.googleapis.com/ajax/libs/jquery/3.5.1/jquery.min.js"></script>
```
Comment (AbhijeetThakur-msft) [inline samples/graph-activity-feed/nodejs/views/GroupChatNotification.ejs:10]: Please check if we need imports for jquery, pooper, etc, and remove if not needed

**Review by AbhijeetThakur-msft — CHANGES_REQUESTED** [review]

File: samples/graph-activity-feed/nodejs/Readme.md (hunk 6 above)
Comment (NikitaJain-MSFT) [reply samples/graph-activity-feed/nodejs/Readme.md:5]: Added sample doc link

File: samples/graph-activity-feed/nodejs/views/auth-start.ejs [hunk 15]
```diff
@@ -0,0 +1,113 @@
...
+        function toQueryString(queryParams) {
+            let encodedQueryParams = [];
+            for (let key in queryParams) {
+                encodedQueryParams.push(key + "=" + encodeURIComponent(queryParams[key]));
+            }
+            return encodedQueryParams.join("&");
+        }
+        // Converts decimal to hex equivalent
+        // (From ADAL.js: https://github.com/AzureAD/azure-activedirectory-library-for-js/blob/dev/lib/adal.js)
+        function _decimalToHex(number) {
```
[the remark this answers is not in the document]
Comment (NikitaJain-MSFT) [reply samples/graph-activity-feed/nodejs/views/auth-start.ejs:45]: this method is used to generate random guid value for state.

**Review by v-chetsh — APPROVED** [review]
All looks good from my end, approving the PR.

**Review by Wajeed-msft — APPROVED** [review]
